spring-cloud-gateway-server-mvc/src/main/java/org/springframework/cloud/gateway/server/mvc/handler/ProxyExchangeHandlerFunction.java

```diff
@@ -84,14 +87,14 @@ private void init() {
 	@Override
 	public ServerResponse handle(ServerRequest serverRequest) {
 		URI uri = uriResolver.apply(serverRequest);
-		boolean encoded = containsEncodedQuery(serverRequest.uri(), serverRequest.params());
+		MultiValueMap<String, String> params = ensureEncodedQueryParameters(serverRequest);
```

Comment:
User can modify the query parameters in preceding filters, for example, add additional parameters to the request, these parameters (`serverRequest.params()`) should be encoded regardless of whether the original request url contains `%` or not.

```suggestion
		MultiValueMap<String, String> params = MvcUtils.encodeQueryParams(serverRequest.params());
```

Reply:
@Shawyeok 

First of all, If query parameters were added in a preceding filter, I believe that filter should be responsible for ensuring those parameters are properly encoded.

`serverRequest.params()` relies on `HttpServletRequest.getParameterMap()`, and from what I can tell, most major servlet containers (Tomcat, Jetty, Undertow) return already-decoded values.

So as you pointed out, it's likely safe to assume that `serverRequest.params()` provides decoded query parameters.

(That said, if we can guarantee that `serverRequest.params()` always returns decoded values, then it's fine to encode them again here. But if that guarantee isn't solid, we risk double-encoding parameters that were already percent-encoded by earlier filters.)

Reply:
> if we can guarantee that serverRequest.params() always returns decoded values, then it's fine to encode them again here.

IMO, the only reliable way to handle URI encoding is to treat the values returned by `serverRequest.params()` as  raw (i.e. decoded values). We're unable to determine whether a string is URI encoded or not, for example: Is `%25` a encoded string? We cannot tell, cause user might intend to send the literal string `%25` rather than the percent symbol. Therefore, we should consistently treat all parameter values as decoded and handle encoding at this stage.